synapseutils/sync.py

```diff
@@ -256,22 +261,39 @@ def sync(self, entity, path, ifcollision, followLink, downloadFile=True, manifes
         files.sort(key=lambda f: f.get('path') or '')
         return files
 
-    def _sync_file(self, entity_id, parent_folder_sync, path, ifcollision, followLink, progress, downloadFile):
+    def _sync_file(self, entity_id, parent_folder_sync, path, ifcollision, followLink, progress, downloadFile, onlyMatching):
         try:
             # we use syn.get to download the File.
             # these context managers ensure that we are using some shared state
             # when conducting that download (shared progress bar, ExecutorService shared
             # by all multi threaded downloads in this sync)
             with progress.accumulate_progress(), \
                     download_shared_executor(self._executor):
-
-                entity = self._syn.get(
-                    entity_id,
-                    downloadLocation=path,
-                    ifcollision=ifcollision,
-                    followLink=followLink,
-                    downloadFile=downloadFile,
-                )
+            
+                file_matches = True
+                if onlyMatching is not None:
+                    file_matches = False
+                    entity_meta = self._syn.get(
+                        entity_id,
+                        downloadFile=False,
+                    )
+                    for regex in onlyMatching:
+                        if re.match(regex, entity_meta.name) is not None:
+                            file_matches = True
+
+                if file_matches:
```

Comment:
I don't think we need the `file_matches` flag. Couldn't we simply short-circuit this function with an early return statement if `onlyMatching` is set and there are no matches? This way, we avoid indenting the `self._syn.get(..., downloadFile=downloadFile)` bit. We would have to run the `parent_folder_sync.update()` statement you have below immediately before the return statement. 

I'll happily elaborate if my suggestion isn't clear.